manager/deallocator/deallocator.go

```diff
@@ -0,0 +1,289 @@
+	})
+
+	if err != nil {
+		log.G(ctx).WithError(err).Errorf("failed to retrieve the list of tasks for service %v", service.ID)
+		// if in doubt, let's proceed to clean up the service anyway
+		// better to clean up resources that shouldn't be cleaned up yet
+		// than ending up with a service and some resources lost in limbo forever
+		return true, deallocator.deallocateService(ctx, service)
+	} else if len(tasks) == 0 {
+		// no tasks remaining for this service, we can clean it up
+		return true, deallocator.deallocateService(ctx, service)
+	}
+	deallocator.services[service.ID] = &serviceWithTaskCounts{service: service, taskCount: len(tasks)}
+	return false, nil
+}
+
+func (deallocator *Deallocator) deallocateService(ctx context.Context, service *api.Service) (err error) {
+	err = deallocator.store.Update(func(tx store.Tx) error {
+		// first, let's delete the service
+		var ignoreServiceID *string
+		if err := store.DeleteService(tx, service.ID); err != nil {
+			// all errors are just for logging here, we do a best effort at cleaning up everything we can
+			log.G(ctx).WithError(err).Errorf("failed to delete service record ID %v", service.ID)
+			ignoreServiceID = &service.ID
+		}
+
+		// then all of its service-level resources, provided no other service uses them
+		for _, resourceDeallocator := range resourceDeallocators {
+			for _, resource := range resourceDeallocator.resourcesExtractor(tx, service) {
+				deallocator.processResource(ctx, tx, resourceDeallocator, resource, ignoreServiceID)
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		log.G(ctx).WithError(err).Errorf("DB error when deallocating service %v", service.ID)
+	}
+	return
+}
+
+// proceeds to deallocating a resource if it's pending deletion and there no
+// longer are any services using it
+func (deallocator *Deallocator) processResource(ctx context.Context, tx store.Tx, resourceDeallocator resourceDeallocator, resource serviceLevelResource, ignoreServiceID *string) (updated bool, err error) {
+	if !resource.isPendingDelete() {
+		return
+	}
+
+	updateFunc := func(t store.Tx) error {
+		services, err := resourceDeallocator.servicesLocator(t, resource.id())
+
+		if err != nil {
+			log.G(ctx).WithError(err).Errorf("could not fetch services using %v ID %v", resourceDeallocator.name, resource.id())
+			return err
+		}
+
+		noMoreServices := len(services) == 0 ||
+			len(services) == 1 && ignoreServiceID != nil && services[0].ID == *ignoreServiceID
+
+		if noMoreServices {
+			return resourceDeallocator.deleter(t, resource.id())
+		}
+		return nil
+	}
+
+	if tx == nil {
+		err = deallocator.store.Update(updateFunc)
+	} else {
+		err = updateFunc(tx)
+	}
+
+	if err != nil {
+		log.G(ctx).WithError(err).Errorf("DB error when deallocating %v ID %v", resourceDeallocator.name, resource.id())
+	}
+	return
+}
+
+func (deallocator *Deallocator) processNewEvent(ctx context.Context, event events.Event) (bool, error) {
+	switch typedEvent := event.(type) {
```

Comment:
lol `typedEvent` is a way better name than `swtich ev := event.(type) {` (which is what we have in other places iirc)